criu/pidfd.c

```diff
@@ -0,0 +1,121 @@
+#include "imgset.h"
+#include "pidfd.h"
+#include "fdinfo.h"
+#include "protobuf.h"
+#include "pstree.h"
+
+#undef LOG_PREFIX
+#define LOG_PREFIX "pidfd: "
+
+struct pidfd_info {
+	PidfdEntry *pidfe;
+	struct file_desc d;
+};
+
+int is_pidfd_link(char *link)
+{
+	/*
+	* pidfs was introduced in Linux 6.9
+	* before which anonymous-inodes were used
+	*/
+	return is_anon_link_type(link, "[pidfd]");
+}
+
+static void pr_info_pidfd(char *action, PidfdEntry *pidfe)
+{
+	pr_info("%s: id %#08x flags %u mnt_id %d ino %u NSpid %u\n",
+		action, pidfe->id, pidfe->flags, pidfe->mnt_id,
+		pidfe->ino, pidfe->nspid
+	);
+}
+
+static int dump_one_pidfd(int pidfd, u32 id, const struct fd_parms *p)
+{
+	PidfdEntry pidfe = PIDFD_ENTRY__INIT;
+	FileEntry fe = FILE_ENTRY__INIT;
+
+	if (parse_fdinfo(pidfd, FD_TYPES__PIDFD, &pidfe))
+		return -1;
+
+	/*
+	* TODO:
+	* First we should check if the process is alive or not
+	* If it is dead:
+	*	dump a ghost pidfd
+	*/
+	/*
+	* Check if the pid pidfd refers to is part of process tree
+	* This ensures the process will exist on restore.
+	*/
+	if (!pstree_item_by_real(pidfe.pid)) {
+		pr_perror("pidfd pid %d is not a part of process tree..", pidfe.pid);
+		return -1;
+	}
+
+	pidfe.id = id;
+	pidfe.flags = (p->flags & ~O_RDWR);
+	pidfe.mnt_id = p->mnt_id;
+	pidfe.fown = (FownEntry *)&p->fown;
+
+	fe.type = FD_TYPES__PIDFD;
+	fe.id = pidfe.id;
+	fe.pidfd = &pidfe;
+
+	pr_info_pidfd("Dumping", &pidfe);
+	return pb_write_one(img_from_set(glob_imgset, CR_FD_FILES), &fe, PB_FILE);
+}
+
+const struct fdtype_ops pidfd_dump_ops = {
+	.type = FD_TYPES__PIDFD,
+	.dump = dump_one_pidfd,
+};
+
+static int pidfd_open(pid_t pid, int flags)
+{
+	return syscall(__NR_pidfd_open, pid, flags);
+}
+
+static int open_one_pidfd(struct file_desc *d, int *new_fd)
+{
+	struct pidfd_info *info;
+	int pidfd;
+
+	info = container_of(d, struct pidfd_info, d);
+	pidfd = pidfd_open(info->pidfe->nspid, info->pidfe->flags);
+	if (pidfd < 0) {
+		goto err_close;
+	}
+
+	if (rst_file_params(pidfd, info->pidfe->fown, info->pidfe->flags)) {
+		goto err_close;
+	}
+
+	*new_fd = pidfd;
+	return 0;
+
+err_close:
+	pr_info_pidfd("Failed to create", info->pidfe);
+	pr_perror("Can't create pidfd %#08x", info->pidfe->id);
```

Comment:
pr_perror should be right after a target call, because any other call between them can over-ride errno.

Reply:
Done!